### ISSUE TITLE
Unresolved reference to firebaseCrashlytics Gradle extension since 2.6.0 [firebase-crashlytics-gradle]

### ISSUE THREAD

**Issue by lordcodes** [firebase/firebase-android-sdk#2665]
### [READ] Step 1: Are you in the right place?

Issues filed here should be about bugs in __the code in this repository__. 
If you have a general question, need help debugging, or fall into some
other category use one of these other channels:

  * For general technical questions, post a question on [StackOverflow](http://stackoverflow.com/)
    with the firebase tag.
  * For general Firebase discussion, use the [firebase-talk](https://groups.google.com/forum/#!forum/firebase-talk)
    google group.
  * For help troubleshooting your application that does not fall under one
    of the above categories, reach out to the personalized
    [Firebase support channel](https://firebase.google.com/support/).

### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: 4.2 (Latest Stable channel)
  * Firebase Component: Firebase Crashlytics Gradle plugin
  * Component version: 2.6.0+ (started in 2.6.0, still happens in latest 2.6.1)

### [REQUIRED] Step 3: Describe the problem

The Gradle configuration to use the 'firebaseCrashlytics' Gradle extension, in order to set configuration properties such as 'mappingFileUploadEnabled', can no longer be resolved. This is using an unchanged set up that worked in previous version and is the same as what is in the Firebase Android documentation.

#### Steps to reproduce:

* Add configuration to app module Gradle configuration to use the 'firebaseCrashlytics' Gradle extension and set any property, such as 'mappingFileUploadEnabled'.
* When you configure the project you get Unresolved errors relating to the extension.

```
e: /src/app/build.gradle.kts:146:13: Unresolved reference: firebaseCrashlytics
e: /src/app/build.gradle.kts:147:17: Unresolved reference: mappingFileUploadEnabled
e: /src/app/build.gradle.kts:164:13: Unresolved reference: firebaseCrashlytics
e: /src/app/build.gradle.kts:165:17: Unresolved reference: mappingFileUploadEnabled
```

If you rollback to the previous version, 2.5.2, everything works again.

#### Relevant Code:

In the app module `build.gradle.kts`:
```
buildTypes {
    getByName("debug") {
        ...

        firebaseCrashlytics {
            mappingFileUploadEnabled = false
        }
    }
    
	...
}
```

**Comment by kmayoral** [comment]
Same issue experienced with previously working 
```
    getByName("debug") {
      ...
      firebaseCrashlytics.mappingFileUploadEnabled = false
    }
    ...
```
example for me as well. 

Android Studio version: tried 4.2 and 4.2.1 
Firebase Component: Firebase Crashlytics Gradle plugin 
Component version: 2.6.1

**Comment by kuFEAR** [comment]
https://github.com/firebase/firebase-android-sdk/issues/2444 same issue

**Comment by kuFEAR** [comment]
also if try use previous generated code (this as ExtensionAware).extensions.getByName("firebaseCrashlytics") as com.google.firebase.crashlytics.buildtools.gradle.CrashlyticsExtension the build failed with error 
```
Extension with name 'firebaseCrashlytics' does not exist. Currently registered extension names: [ext, libraries, libs, versionCatalogs, defaultArtifacts, sourceSets, reporting, java, javaToolchains, buildOutputs, android, androidComponents, room, kotlin, kotlinTestRegistry, androidExtensions, kapt, googleServices]
```

**Comment by mrichards** [comment]
Hi folks, we're aware of the regression and plan on improving support for the Gradle Kotlin DSL soon, as adoption seems to be picking up across the Android ecosystem. For now, I believe the `withGroovyBuilder` workaround described in #2444 still works.

**Comment by mrichards** [comment]
@lordcodes @kmayoral In addition to the `withGroovyBuilder` workaround, I believe use of the Kotlin DSLs [ExtensionAware.the()](https://gradle.github.io/kotlin-dsl-docs/api/org.gradle.kotlin.dsl/org.gradle.api.plugins.-extension-aware/the.html) functions will work:

```
import com.google.firebase.crashlytics.buildtools.gradle.CrashlyticsExtension
// ...
 getByName("debug") {
      // ...
      the<CrashlyticsExtension>().mappingFileUploadEnabled = false
      // or
      the(CrashlyticsExtension::class).mappingFileUploadEnabled = false
}
```

This is a little less attractive as it exposes the `CrashlyticsExtension` class in your build script rather than allowing you to use the `firebaseCrashlytics` block, but it gives better type safety than the `withGroovyBuilder` workaround. Thoughts on this?

Just be advised that technically `CrashlyticsExtension` isn't part of our public API. However, if we break this syntax it will be while adding improved Kotlin support so we'll be sure to publish official docs describing the preferred approach.

Finally, can either of you confirm that using the `firebaseCrashlytics.mappingFileUploadEnabled = false` syntax actually worked as expected in a Kotlin build script? In my test apps using 2.5.2, my builds succeed but the value isn't respected.

**Comment by NathanMDT** [comment]
Also experiencing it. 
This documentation has not been changed. 
https://firebase.google.com/docs/crashlytics/get-deobfuscated-reports?platform=android
But following the documentation produces the error.

**Comment by lordcodes** [comment]
> @lordcodes @kmayoral In addition to the `withGroovyBuilder` workaround, I believe use of the Kotlin DSLs [ExtensionAware.the()](https://gradle.github.io/kotlin-dsl-docs/api/org.gradle.kotlin.dsl/org.gradle.api.plugins.-extension-aware/the.html) functions will work:

Thank you, I will try that out when I get the chance. Currently been stuck on 2.5.2 and would like to free myself up to update the plugin further.

> Finally, can either of you confirm that using the `firebaseCrashlytics.mappingFileUploadEnabled = false` syntax actually worked as expected in a Kotlin build script? In my test apps using 2.5.2, my builds succeed but the value isn't respected.

It is in the Firebase Android documentation, specifically supplying Kotlin DSL sample code, so based on that I assumed it was working. I will have a look into this too, when I get the chance.

---

Personally, I think the documentation needs updating so that it has the correct code in it for the latest version of plugin. Also, it seems like maybe the Kotlin Gradle DSL isn't considered a priority, if it was broken entirely in the 2.6.0 release, from your phrasing it sounds like fixing the Kotlin DSL is seen more as an "improvement". Therefore, a statement in the documentation that says the Kotlin DSL is considered experimental might be a good idea as well.

Anyway, thank you very much for the help. I will implement your suggestion so that I can update to 2.6.0 again!

**Comment by mrichards** [comment]
You are correct that the Kotlin Gradle DSL hasn't been a big priority for the Crashlytics team, though we're working on improving it and will be adding official documentation soon. In the meantime, _it is likely_ that the following syntax is going to be the one we recommend moving forward:

```
import com.google.firebase.crashlytics.buildtools.gradle.CrashlyticsExtension

android {
  buildTypes{
        getByName("release") {
            configure<CrashlyticsExtension> {
                nativeSymbolUploadEnabled = true      // to enable NDK symbol file uploading (default=false)
                mappingFileUploadEnabled = false       // to disable mapping file uploads (default=true if minifying)
            }
        }
    }
}
```

This will work as expected for v2.6.0+ of the Crashlytics plugin, and is a little cleaner (IMO) than using the `the` method which I described above. Using `firebaseCrashlytics.mappingFileUploadEnabled = false` in the Kotlin DSL pre-2.6.0 was not reliable, which was an unknown bug until recently. Using that syntax with the Kotlin DSL in 2.6.0+ now results in a build failure, as several of you found out.

We added a Kotlin DSL build to our test matrix for the Crashlytics plugin. Any breaking changes the the `configure<>` syntax above will, at minimum, be noted in the release notes and accompanied by a major version increase.
